store/src/java/com/zimbra/cs/service/mail/ToXML.java

```diff
@@ -1631,7 +1621,36 @@ private static Element encodeMessageAsMPHelper(boolean bestEffort, Element paren
             List<MPartInfo> parts = Mime.getParts(mm, getDefaultCharset(msg));
             if (parts != null && !parts.isEmpty()) {
                 Set<MPartInfo> bodies = Mime.getBody(parts, wantHTML);
-                addParts(m, parts.get(0), bodies, part, maxSize, neuter, false, getDefaultCharset(msg), bestEffort, wantContent);
+                // added condition to check if email is external and if it contains warning in html content
```

Comment:
I think we already have code at lmtp or smtp level to add warning message based on zimbra_external_email_warning_message value, so isn't that code working here? and why we are adding completely new code instead of fixing that?

Reply:
yes, we have code to add external warning to incoming email. It is working as expected for plain text but failing for html content for Emails coming from gmail domain. 

Current code changes are on presentation layer wherein below conditions are checked -
1.  If Email is external
2. If html content contains base warning(extracted from plain text content)

if the warning is missing in html content we are transforming plain text warning to html warning and adding it to response.

Reply:
can't we fix code to add warning at LMTP/SMTP level for html mails also?

Reply:
Consider scenario:
1) currently admin has not added a warning attribute
2) user gets some mails
3) admin adds new warning attribute
4) if user opens earlier mails it will add warning, I guess should not be done

Reply:
In such scenarios default warning will be display.

For eg.

1. Admin has not set the warning and receives external email (**M1**). In such scenario default warning will be used (default warning ~ This message originated outside of your organization.) and email will be stored.
2. Admin Edits the warning msg and receives another external email. In this scenario edited warning will be used and mail will be stored.
3. If the User tries to access M1 later, it will show the default warning as it was warning at the time of processing/storage.

Warning methods at LTMP/SMTP are invoked only when a new email is received, so if we make changes at LMTP/SMTP layers, only new emails will contain warning in both plain & html content and issue will still persist for existing email. To avoid this, changes are made at presentation layer where response is created.

Reply:
I still feel, we should fix this at LMTP/SMTP layer only and not at presentation layer.